ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll

```diff
@@ -264,3 +465,195 @@ private DataFlowPublic::Node evaluateSummaryComponentLocal(
+    or
+    callable.propagatesFlow(_, stack, true)
+  )
+  or
+  dependsOnSummaryComponentStackCons(callable, _, stack)
+}
+
+pragma[nomagic]
+private predicate dependsOnSummaryComponentStackCons(
+  SummarizedCallable callable, SummaryComponent head, SummaryComponentStack tail
+) {
+  dependsOnSummaryComponentStack(callable, SCS::push(head, tail))
+}
+
+pragma[nomagic]
+private predicate dependsOnSummaryComponentStackConsLocal(
+  SummarizedCallable callable, SummaryComponent head, SummaryComponentStack tail
+) {
+  dependsOnSummaryComponentStackCons(callable, head, tail) and
+  not isNonLocal(head)
+}
+
+pragma[nomagic]
+private predicate dependsOnSummaryComponentStackLeaf(
+  SummarizedCallable callable, SummaryComponent leaf
+) {
+  dependsOnSummaryComponentStack(callable, SCS::singleton(leaf))
+}
+
+/**
+ * Gets a data flow node corresponding to the local input or output of `call`
+ * identified by `stack`, if possible.
+ */
+pragma[nomagic]
+private DataFlow::Node evaluateSummaryComponentStackLocal(
+  SummarizedCallable callable, DataFlow::CallNode call, SummaryComponentStack stack
+) {
+  exists(SummaryComponent component |
+    dependsOnSummaryComponentStackLeaf(callable, component) and
+    stack = SCS::singleton(component) and
+    call.asExpr().getExpr() = callable.getACallSimple() and
+    result = evaluateSummaryComponentLocal(call, component)
+  )
+  or
+  exists(DataFlow::Node prev, SummaryComponent head, SummaryComponentStack tail |
+    prev = evaluateSummaryComponentStackLocal(callable, call, tail) and
+    dependsOnSummaryComponentStackConsLocal(callable, pragma[only_bind_into](head),
+      pragma[only_bind_out](tail)) and
+    stack = SCS::push(pragma[only_bind_out](head), pragma[only_bind_out](tail))
+  |
+    exists(DataFlowDispatch::ArgumentPosition apos, DataFlowDispatch::ParameterPosition ppos |
+      head = SummaryComponent::parameter(apos) and
+      DataFlowDispatch::parameterMatch(ppos, apos) and
+      result.(DataFlowPrivate::ParameterNodeImpl).isSourceParameterOf(prev.asExpr().getExpr(), ppos)
+    )
+    or
+    head = SummaryComponent::return() and
+    result.(DataFlowPrivate::SynthReturnNode).getCfgScope() = prev.asExpr().getExpr()
+    or
+    exists(DataFlow::ContentSet content |
+      head = SummaryComponent::withoutContent(content) and
+      not exists(getFilterFromWithoutContentStep(content)) and
+      result = prev
+    )
+  )
+}
+
+private newtype TUnit = MkUnit()
+
+/**
+ * A data flow edge that should be followed by type tracking.
+ *
+ * This type of edge does not affect the local data flow graph, and is not used by data-flow configurations.
+ *
+ * Note: For performance reasons, all subclasses of this class should be part
+ * of the standard library, and their implementations may not depend on API graphs.
+ * For query-specific steps, consider including the custom steps in the type-tracking predicate itself.
+ */
+class TypeTrackingStep extends TUnit {
```

Comment:
I am not convinced that we should add this, especially as it is not used presently.